### PR TITLE
IMU Code and Changes to Serial Manager

### PR DESCRIPTION
**Pull request by lymacasm**
Completed code for #53 and #105. While testing, changes were made to the serial_manager to get it working. A new feature was also added to the serial manager that allows us to ignore certain peripherals while testing. IMU node publishes to /imu/imu_sensor with IMU readings along with calculated velocity. The velocity is calculated by integrating the (instantaneous acceleration - stabilized acceleration) using the trapezoid integration method. The IMU updates its velocity faster than it publishes to the imu_sensor topic so that it can crunch the numbers at a higher rate. IMU launch file has been added with the ability to record bags, and the imu launch as well as the serialmanager launch files have been referenced in local.launch. Currently publishing more information than we need for navigation, could be useful for data logging.

### CODE REVIEW

**Comment by lymacasm** [comment]
It hasn't been rebased, as you can do that very nicely using the merge button on this site. See "Squash and merge" in dropdown. (One commit, with all commit messages included)

File: ros/src/peripherals/launchfiles/imu.launch [hunk 1]
```diff
@@ -1,6 +1,22 @@
...
+
+    <node name="$(arg node_name)" pkg="peripherals" type="imu" respawn="true" respawn_delay="1" >
+        <param name="topic_name" value="$(arg topic_name)"/>
+        <param name="topic_buffer_size" value="$(arg topic_buffer_size)"/>
         <param name="device_id" value="$(arg device_id)"/>
+        <param name="publish_rate" value="$(arg publish_rate)"/>
+        <param name="updates_per_publish" value="$(arg updates_per_publish)"/>
     </node>
+
+    <group if="$(arg record)">
```
Comment (asinha94) [inline ros/src/peripherals/launchfiles/imu.launch:19]: Im not sure if we want individual nodes to record to rosbag over just a single rosbag which records everything. Ill leave this for now

File: ros/src/monitor/src/serial_manager.cpp [hunk 2]
```diff
@@ -25,12 +30,13 @@ device_manager::device_manager(const std::vector<device_property> & properties)
                 continue;
             }
             
-            serial::Serial connection(port->port, (uint32_t) property->baud , property->timeout);
+            std::unique_ptr<serial::Serial> connection = std::unique_ptr<serial::Serial>(
```
Comment (asinha94) [inline ros/src/monitor/src/serial_manager.cpp:33]: Any reason why we need to make the pointer over the raw object? I know I usually use the pointer versions when I embed them into a class in some of the other node, but dont think it makes a difference

**Review by asinha94 — APPROVED** [review]
Looks good. My comments are just curiosities more than anything else. Ill let you squash and merge so that you can get onto the contributors scoreboard

File: ros/src/monitor/src/serial_manager.cpp (hunk 2 above)
Comment (lymacasm) [reply ros/src/monitor/src/serial_manager.cpp:33]: This caused a strange bug in the serial_manager. Reading wasn't working properly (garbage values were returned). After switching to the pointer, reading worked fine and I was able to find the IMU device.

File: ros/src/peripherals/launchfiles/imu.launch (hunk 1 above)
Comment (lymacasm) [reply ros/src/peripherals/launchfiles/imu.launch:19]: Good point. Maybe we could have a peripherals launch file that optionally recorded all peripherals into a bag file?

**Comment by lymacasm** [comment]
Awesome, thanks Anu!

**Pull request merged**